packages/react-query/src/shared/hooks/createHooksInternal.tsx

```diff
@@ -181,7 +182,36 @@ export function createRootHooks<
                 },
               };
 
-              return client.query(...getClientArgs(queryKey, actualOpts));
+              const result = await client.query(
+                ...getClientArgs(queryKey, actualOpts),
+              );
+
+              if (isAsyncIterable(result)) {
+                const queryHash = hashKey(queryKey);
+
+                const queryCache = queryClient.getQueryCache();
+
+                const query = queryCache.build(queryClient, {
+                  queryKey,
+                  queryHash,
+                });
+
+                query.setState({
+                  data: [],
+                  status: 'pending',
+                });
```

Comment:
Would love TkDodo's opinion on this. When I was trying to think of this DX I couldn't decide between the possible "is streaming" states
- `status: 'pending', fetchStatus: 'fetching', data: <defined>`
- `status: 'success', fetchStatus: 'fetching', data: <defined>`
- `status: 'success', fetchStatus: 'idle', data: <defined>`

Reply:
Hmm you might have a point - I rarely think of fetchStatus

Reply:
cc @TkDodo wdyt is best here?

Reply:
in `pending` state, data should be `undefined`. That's how we have it typed, `pending` means "no data": https://github.com/TanStack/query/blob/8f55e36a7c94242176875bbb18c63ab6991c6af0/packages/query-core/src/types.ts#L651-L658

if `data` is defined, this is probably the most correct representation:

```
status: 'success', fetchStatus: 'fetching', data: <defined>
```